pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +449,705 @@ pub mod pallet {
+					self.decrease_bottom_delegation::<T>(candidate, delegator, less)
+				} else {
+					result
+				}
+			} else {
+				self.decrease_bottom_delegation::<T>(candidate, delegator, less)
 			}
 		}
-		/// Return true if in_top after call
-		pub fn decrease_delegation(&mut self, delegator: A, less: B) -> bool {
-			let mut in_top = false;
-			let mut new_lowest_top: Option<Bond<A, B>> = None;
-			for x in &mut self.top_delegations {
-				if x.owner == delegator {
-					x.amount -= less;
-					// if there is at least 1 delegator in bottom delegators, compare it to check
-					// if it should be swapped with lowest top delegation and put in top
-					// >pop requires push later on to reset in case it isn't used
-					if let Some(highest_bottom) = self.bottom_delegations.pop() {
-						if highest_bottom.amount > x.amount {
-							new_lowest_top = Some(highest_bottom);
-						} else {
-							// >required push to reset self.bottom_delegations
-							self.bottom_delegations.push(highest_bottom);
+		/// Decrease top delegation
+		pub fn decrease_top_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+			bond: Balance,
+			less: BalanceOf<T>,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let bond_after_less_than_highest_bottom =
+				bond - less.into() < self.highest_bottom_delegation_amount;
+			let full_top_and_nonempty_bottom = matches!(self.top_capacity, CapacityStatus::Full)
+				&& !matches!(self.bottom_capacity, CapacityStatus::Empty);
+			let mut top_delegations =
+				<TopDelegations<T>>::get(candidate).ok_or(Error::<T>::CandidateDNE)?;
+			let in_top_after =
+				if bond_after_less_than_highest_bottom && full_top_and_nonempty_bottom {
+					let mut delegation_option: Option<Bond<T::AccountId, BalanceOf<T>>> = None;
+					// take delegation from top
+					top_delegations.delegations = top_delegations
+						.delegations
+						.clone()
+						.into_iter()
+						.filter_map(|d| {
+							if d.owner != delegator {
+								Some(d)
+							} else {
+								let new_amount = d.amount.saturating_sub(less);
+								top_delegations.total -= d.amount;
+								delegation_option = Some(Bond {
+									owner: d.owner,
+									amount: new_amount,
+								});
+								None
+							}
+						})
+						.collect();
+					let delegation = delegation_option.ok_or(Error::<T>::DelegationDNE)?;
+					// pop highest bottom by reverse and popping
+					let mut bottom_delegations = <BottomDelegations<T>>::get(candidate)
+						.expect("CandidateInfo existence => BottomDelegations existence");
+					let highest_bottom_delegation = bottom_delegations.delegations.remove(0);
+					bottom_delegations.total -= highest_bottom_delegation.amount;
+					// insert highest bottom into top
+					top_delegations.insert_sorted_greatest_to_least(highest_bottom_delegation);
+					// insert previous top into bottom
+					bottom_delegations.insert_sorted_greatest_to_least(delegation);
```

Comment:
This should always be going into the top, right?

In fact, even if there is a tie for the top, you want to make sure it goes at the beginning of the identical bonds -- not the end (otherwise we fail to preserve insertion order fairness) -- right?

Reply:
The highest bottom is going into the top and the changed delegation is going into the bottom.

The condition for this branch to execute is `bond_after_less_than_highest_bottom` and it is a strict less than so we definitely want to insert the decreased delegation into the bottom and pop the highest bottom into the top.